### PR TITLE
[MRG] Fix missing comps in CTF data.

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/4108

This only fixes missing columns. I didn't want to change the rows as I don't quite understand how it is possible to have these missing in the first place.

### CODE REVIEW

**Comment by agramfort** [comment]
can you see how fieldtrip or brainstorm handle this?

maybe @robertoostenveld knows... see #4108

**Comment by Akshay0724** [comment]
After going through the FieldTrip codebase I found that in case of CTF data the `grad` structure included in fieldtrip contain a key as `balance` to make it consistent with the data.

Generally `grad.balance` look like this -

```
 G1BR: [1x1 struct]   // first order balancing
 G2BR: [1x1 struct]   // second order balancing
 G3BR: [1x1 struct]   // third order balancing
 current: 'none'         // no balancing
```
But when we use ArtifactMEG dataset than `G3BR` field is not present, instead this message is shown-

```
getRawCTFBalanceCoefs : Sensor Q11 appears in ds.res4.scrr, but not in ds.res4.chanNames

Warning: cannot read balancing coefficients for G3BR 
> In ft_read_header (line 556) 
```

I think instead of discarding whole balancing scheme, we should remove the culprit channel as done in this PR.

Also a warning must be shown indicating the problem and the corresponding action taken because it's a problem with dataset not with the reader.

**Comment by agramfort** [comment]
thanks for taking a close look !

can you check/compare the outputs of the readers using this branch and
fieldtrip?

**Comment by Akshay0724** [comment]
hi **@agramfort**,
please conform did you mean to check the output raw,epoch,evoked data.

Thanks

**Comment by agramfort** [comment]
I meant check the numerical values after reading the file. What you get by
doing raw[:, :]

**Comment by Akshay0724** [comment]
Sorry for late reply. I tried to compare the raw data but found the cardinality is not same  in both places.

Data in filedtrip comes divided into trials and have form `n_trial X n_channel X trial_time` and for _ArtifactMEG.ds_ data it is 76 * 176 * 12000 equivalent to 176 * 912,000 in 2D.

But in MNE we are having 176 * 900,000 raw data. The difference is exactly equal to one trial of length 12000.

 On the other hand I had not found any issue with other data set like _Subject01.ds_ of fieldtrip. I'm trying to figure out the actual cause but it would be excellent if someone can provide any pointers.

**Comment by jaeilepp** [comment]
It's probably because of the dropped comps. I'll look into it.

**Comment by jaeilepp** [comment]
Looks like this has nothing to do with the comps. It actually looks like a bug to me. In effect, the last block of data is almost always dropped in mne. The reader looks at the end of the data and cuts of at the last data sample, but then discards the last block.

File: mne/io/ctf/ctf.py [hunk 1]
```diff
@@ -214,14 +214,7 @@ def _get_sample_info(fname, res4, system_clock):
...
         logger.info('    %d x %d = %d samples from %d chs'
                     % (n_trial, n_samp, n_samp, res4['nchan']))
-    else:
-        n_trial = n_samp // res4['nsamp']
-        n_omit = n_samp_tot - n_samp
-        n_samp = n_trial * res4['nsamp']
-        logger.info('    %d x %d = %d samples from %d chs'
-                    % (n_trial, res4['nsamp'], n_samp, res4['nchan']))
-        if n_omit != 0:
-            logger.info('    %d samples omitted at the end' % n_omit)
```
Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: I wonder what was the reasoning behind this.

**Comment by robertoostenveld** [comment]
Data is written in blocks (by default 10 seconds) but acquisition can be stopped at any moment in time. I believe that in the CTF acquisition software the last block is then zero-padded. So the number of valid samples might be smaller than the total number of samples (which is a multiple of blocks).

**Comment by jaeilepp** [comment]
Yeah, probably it was trying to just cut the zeros, but throws away the end block by accident. I'll put the logging back.

File: mne/io/ctf/ctf.py [hunk 2]
```diff
@@ -217,11 +217,11 @@ def _get_sample_info(fname, res4, system_clock):
     else:
         n_trial = n_samp // res4['nsamp']
         n_omit = n_samp_tot - n_samp
-        n_samp = n_trial * res4['nsamp']
```
Comment (Eric89GXL) [inline mne/io/ctf/ctf.py]: I suspect that breaking this relationship could eventually break some code downstream. Can you try getting rid of `n_trial` and `n_samp_tot` from the output dict, and see if the downstream code still works or can be refactored to work? That would show places where breaking this relationship could be problematic.

Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: I don't see ``n_trial`` used anywhere after this, and ``n_samp_tot`` is only used to [determine the size on disk](https://github.com/mne-tools/mne-python/blob/master/mne/io/ctf/ctf.py#L155).

Comment (Eric89GXL) [inline mne/io/ctf/ctf.py]: That line should probably change to just use `n_samp` now, right? It seems like `n_samp_tot` isn't actually the size on disk, unless the zeros actually get written to disk.

Comment (Eric89GXL) [inline mne/io/ctf/ctf.py]: (And we might as well remove `n_trial` from the dict, then.)

Comment (jaeilepp) [inline mne/io/ctf/ctf.py]: I believe the zeros get written on the disk. This also works if ``n_samps`` is set to ``n_samps_tot``, but then the data would contain zeros at the end.

Comment (Eric89GXL) [inline mne/io/ctf/ctf.py]: Ahh okay, makes sense. Then let's get rid of `n_trial` from the dict and we should be good.

**Comment by Eric89GXL** [comment]
@Akshay0724 can you see if this works for you now?

**Comment by Akshay0724** [comment]
Now I'm getting `176 x 902106` but according to fieldtrip this should be `176 x 912000`. This cardinality issue only occurs when missing channels are present in comps while other datasets gives same result.

**Comment by jaeilepp** [comment]
Data at the end in fieldtrip is probably all zeros. The data points end at 902106.

**Comment by jaeilepp** [comment]
I'm changing this to MRG as I don't think this breaks anything and it includes a bug fix.

**Comment by agramfort** [comment]
thx @jaeilepp

**Pull request merged**